nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py

```diff
@@ -875,6 +877,10 @@ def get_parameters(self):
                 params.append(param)
         return params
 
+    def generate(self, inputs: Union[List[str], Tensor, List[dict]], sampling_params: SamplingParam) -> OutputType:
+        # no implementation
+        return super().generate(inputs, sampling_params)
+
     def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
```

Comment:
I'm wondering how will we maximize the code re-use between predict/eval/test_step and .generate?
Will you be relying on ._bucketize_gpt_inference under the hood of ".generate"?

Reply:
I would do the opposite. The `.generate` calls `trainer.predict` as we need it to initialize the DDPPlugin. So `predict_step` and `bucket` For `eval` and `test`, it just evaluate the LM loss which is different from text generation.  
` ._bucketize_gpt_inference` will not be used as the `text_generation_util` padded the tokens into the same length.

Reply:
what if Model was created without passing the trainer instance (non-model parallel)?

Reply:
Even non-model parallel models require Trainer.

Reply:
@ericharper I think it can be None for non-model parallel stuff, isn't it https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/nlp_model.py#L48 ?

Reply:
for NLPModels, but megatron layers require ddp

Reply:
https://github.com/NVIDIA/NeMo/blob/c4ec081c2a6d718f8917155d55a006b5f6fcbb27/nemo/collections/nlp/modules/common/megatron/transformer.py#L689